Only accept string literals as translation macro arguments

lanurmi opened the pull request:
Point translation macros to wrapper functions that only accept string literals. This is a backward-incompatible change, though most incompatibilities found are likely to be bugs.

While at present the arguments do not technically need to be string literals, variables can easily introduce subtle bugs that are not obvious to spot. The fact that xgettext only picks string literals is easy to forget.

Consider  the following code:
```c++
wxString itm("item");
_(itm);
wxGETTEXT_IN_CONTEXT("context_1", itm);
wxGETTEXT_IN_CONTEXT(itm, "item");
wxPLURAL(wxString::Format("%s", "singX"), "plurX", 1);
wxPLURAL("singY", wxString::Format("%s", "plurY"), 2);
wxGETTEXT_IN_CONTEXT_PLURAL("context_1", _("singZ"), "plurZ", 1);
```

All of these lines compile without warnings. However, only one string, `"singZ"`, is picked by `xgettext`, all others are silently ignored. Some lines would work if accompanied by a proper `wxTRANSLATE*` call, but nested calls (see the last line) to translation macros do not make sense in the first place.

lanurmi committed:
Only accept string literals as translation macro arguments

Point translation macros to wrapper functions that only accept
string literals. This is a backward-incompatible change,
though most incompatibilities found are likely to be bugs.

While at present the arguments do not technically need to be string
literals, variables can easily introduce subtle bugs that are not
obvious to spot. The fact that xgettext only picks string literals
is easy to forget.

Consider  the following code:
wxString itm("item");
_(itm);
wxGETTEXT_IN_CONTEXT("context_1", itm);
wxGETTEXT_IN_CONTEXT(itm, "item");
wxPLURAL(wxString::Format("%s", "singX"), "plurX", 1);
wxPLURAL("singY", wxString::Format("%s", "plurY"), 2);
wxGETTEXT_IN_CONTEXT_PLURAL("context_1", _("singZ"), "plurZ", 1);

All of these lines compile without warnings. However, only one string,
"singZ", is picked by xgettext, all others are silently ignored. Some
lines would work if accompanied by a proper wxTRANSLATE* call, but nested
calls (see the last line) to translation macros do not make sense in the
first place.

vadz commented on include/wx/translation.h:
```diff
@@ -45,30 +45,19 @@ using wxTranslationsHashMap = std::unordered_map<wxString, wxString>;
-#else
-    #define _(s)                               wxGetTranslation(wxASCII_STR(s))
-#endif
+    #define _(s)                               wx_Wrapper((s))
```
Pretty minor, but could we call it something more readable, e.g.

```suggestion
    #define _(s)                               wxUnderscoreWrapper((s))
```

?

vadz reviewed (commented):
Thanks, this definitely looks like a good change, but I'm afraid we need to provide some way of allowing to disable this as I suspect that there is a lot of code which is not going to compile with it (and yes, most probably this will be due to i18n bugs in it, but i18n bugs are not as bad as failing to compile at all).

To make this more manageable, and also to avoid the pretty weird "wxwx" suffixes, what about defining the new functions in some `wxTransImplStrict` namespace and then conditionally using or not using it depending on some `wxNO_ONLY_LITERAL_TRANSLATIONS` or something similar?

lanurmi committed:
Document that translation macros require literal arguments

While this is technically a change in 3.3.0, the macros always required
string literal arguments in practice, although it was not enforced.

lanurmi commented:
> To make this more manageable, and also to avoid the pretty weird "wxwx" suffixes, what about defining the new functions in some `wxTransImplStrict` namespace

I'll do that.

lanurmi committed:
Optionally re-allow passing variables to translation macros

Defining wxNO_REQUIRE_LITERAL_MSGIDS enables the old behaviour, where
translation macros do accept variables as string arguments.

The arguments nearly always need to be string literals, though, and
before defining this macro, it is advisable to check whether the code
should be fixed instead.

lanurmi committed:
fixup! Optionally re-allow passing variables to translation macros

lanurmi committed:
fixup! Only accept string literals as translation macro arguments

vadz commented on include/wx/translation.h:
```diff
@@ -346,6 +335,116 @@ inline const wxString& wxGetTranslation(const char *str1,
+// Wrapper functions that only accept string literals as arguments,
+// not variables, not char* pointers.
+template<size_t N>
+const wxString &wxUnderscoreWrapper(const char (&msg)[N])
```
Extremely minor style nitpick, but could we please use this style:

```suggestion
const wxString& wxUnderscoreWrapper(const char (&msg)[N])
```

everywhere here?

vadz reviewed (commented):
Thanks a lot! This looks much nicer now but, in addition to some real nitpicks, I'd like to make it even better by getting rid of `wxNO_IMPLICIT_WXSTRING_ENCODING` checks, which should be simple to do, unless I'm missing something — please let me know if I do.

Thanks again!

lanurmi committed:
fixup! fixup! Only accept string literals as translation macro arguments

lanurmi committed:
fixup! fixup! Optionally re-allow passing variables to translation macros

lanurmi committed:
fixup! fixup! fixup! Optionally re-allow passing variables to translation macros

lanurmi committed:
fixup! Document that translation macros require literal arguments

lanurmi committed:
fixup! fixup! Document that translation macros require literal arguments

lanurmi commented on interface/wx/translation.h:
```diff
@@ -465,9 +465,11 @@ class wxMsgCatalog
+    This macro is similar to _() but for the plural variant of
     wxGetTranslation().
 
+    The string arguments must be @a string @a literals.
```
Yes, indeed, `@em` was meant, but `@a` just blindly copied from something that was in italics.

lanurmi commented on interface/wx/translation.h:
```diff
@@ -567,7 +573,8 @@ class wxMsgCatalog
     also common in Unix world) syntax is provided: the _() macro is defined to
-    do the same thing as wxGetTranslation().
+    do nearly the same thing as wxGetTranslation(), with the exception that
+    arguments to _() must be string literals.
```
Will be changed. As for why: the idea supposedly was to say "arguments to [all the] _() [calls, anywhere] must be string literals". But maybe it isn't completely grammatical.

lanurmi committed:
Use macro to avoid repetitive #ifndefs for wxASCII_STR

vadz commented:
Thanks again for the fixes and the changes, will merge in a moment.

vadz commented:
Sorry, I think there is a problem here: the "plural" argument of `wxPLURAL` is never actually used to look up the translation, so it has to be a `wxString` and can't be a string literal. E.g. I think something like this:

```cpp
                            wxPLURAL
                            (
                              "one modified object",
                              wxString::Format("%d modified objects", countModified),
                              countModified
                            ),
```

is correct, but it is flagged as an error now. Am I missing something or does this need to be fixed?

lanurmi commented:
> Sorry, I think there is a problem here: the "plural" argument of `wxPLURAL` is never actually used to look up the translation,

Yes, it looks like that only the "singular" argument is used for translation lookup when loading them, but that doesn't make the "plural" argument irrelevant to the process of extracting translations.

> so it has to be a `wxString` and can't be a string literal.

Sorry, but it shouldn't be a `wxString` and must be a string literal. And while earlier it wasn't enforced to be a string literal, it should've been one then, too.

> E.g. I think something like this:
> 
> ```c++
>                             wxPLURAL
>                             (
>                               "one modified object",
>                               wxString::Format("%d modified objects", countModified),
>                               countModified
>                             ),
> ```

Your `wxPLURAL` expression would not be extracted by `xgettext`, because `xgettext` expects to see two string literals. Seeing `wxString::Format` between them makes it ignore the expression. So this message would not be included in the translation file, and consequently nobody would translate it.

Even if `xgettext` would extract it, and ignoring the fact that `"%d modified objects"` isn't marked for translation either, what does `wxPLURAL` actually return in this case? If it looked up a translation for formatted e.g. `"5 modified objects"`, literally, that would not be found. And if indeed only the "singular" argument is used for lookup, then I guess it will return the _unformatted_ string, e.g. `"%d objetos modificados"`, and the result of the `Format` would be thrown away.

> is correct, but it is flagged as an error now. Am I missing something or does this need to be fixed?

Your code sample nicely demonstrates exactly the kind of bug the improvements of this PR aimed to catch in the first place.

The call really should be like this:
```c++
    wxString::Format(wxPLURAL
        (
            "one modified object",
            "%d modified objects", countModified
        ),
        countModified
    ),
```
